File: packages/runar-py/runar/sdk/woc_provider.py
"""WhatsOnChainProvider — HTTP-based BSV API provider.

Uses only stdlib (urllib.request) for HTTP — no external dependencies required.
"""

from __future__ import annotations

import json
import math
from urllib.error import HTTPError
from urllib.request import Request, urlopen

from runar.sdk.provider import Provider
from runar.sdk.types import TransactionData, TxInput, TxOutput, Utxo


class WhatsOnChainProvider(Provider):
    """Implements Provider using the WhatsOnChain REST API."""

    def __init__(self, network: str = 'mainnet'):
        self.network = network
        if network == 'mainnet':
            self.base_url = 'https://api.whatsonchain.com/v1/bsv/main'
        else:
            self.base_url = 'https://api.whatsonchain.com/v1/bsv/test'

    def _get(self, path: str) -> bytes:
        """Perform a GET request and return the response body bytes."""
        url = f'{self.base_url}{path}'
        req = Request(url, method='GET')
        try:
            resp = urlopen(req, timeout=30)
            return resp.read()
        except HTTPError as e:
            body = e.read()
            raise RuntimeError(
                f'WoC GET {path} failed ({e.code}): {body.decode("utf-8", errors="replace")}'
            ) from e

    def _post_json(self, path: str, data: dict) -> bytes:
        """Perform a POST request with JSON body and return the response body bytes."""
        url = f'{self.base_url}{path}'
        body = json.dumps(data).encode('utf-8')
        req = Request(
            url,
            data=body,
            method='POST',
            headers={'Content-Type': 'application/json'},
        )
        try:
            resp = urlopen(req, timeout=30)
            return resp.read()
        except HTTPError as e:
            err_body = e.read()
            raise RuntimeError(
                f'WoC POST {path} failed ({e.code}): {err_body.decode("utf-8", errors="replace")}'
            ) from e

    def get_transaction(self, txid: str) -> TransactionData:
        raw = self._get(f'/tx/hash/{txid}')
        data = json.loads(raw)

        inputs: list[TxInput] = []
        for vin in data.get('vin', []):
            inputs.append(TxInput(
                txid=vin['txid'],
                output_index=vin['vout'],
                script=vin.get('scriptSig', {}).get('hex', ''),
                sequence=vin.get('sequence', 0xFFFFFFFF),
            ))

        outputs: list[TxOutput] = []
        for vout in data.get('vout', []):
            satoshis = round(vout['value'] * 1e8)
            script_hex = vout.get('scriptPubKey', {}).get('hex', '')
            outputs.append(TxOutput(satoshis=satoshis, script=script_hex))

        return TransactionData(
            txid=data['txid'],
            version=data.get('version', 1),
            inputs=inputs,
            outputs=outputs,
            locktime=data.get('locktime', 0),
            raw=data.get('hex', ''),
        )

    def broadcast(self, tx) -> str:
        # Accept either a raw hex string or an object with .hex() method
        if isinstance(tx, str):
            raw_tx = tx
        else:
            raw_tx = tx.hex()

        resp_body = self._post_json('/tx/raw', {'txhex': raw_tx})
        # WoC returns the txid as a JSON-encoded string
        txid = json.loads(resp_body)
        if isinstance(txid, str):
            return txid
        return str(txid)

    def get_utxos(self, address: str) -> list[Utxo]:
        raw = self._get(f'/address/{address}/unspent')
        entries = json.loads(raw)

        utxos: list[Utxo] = []
        for e in entries:
            utxos.append(Utxo(
                txid=e['tx_hash'],
                output_index=e['tx_pos'],
                satoshis=e['value'],
                script='',  # WoC doesn't return locking script in UTXO list
            ))
        return utxos

    def get_contract_utxo(self, script_hash: str) -> Utxo | None:
        try:
            raw = self._get(f'/script/{script_hash}/unspent')
        except RuntimeError as e:
            # 404 means no UTXO found
            if '404' in str(e):
                return None
            raise

        entries = json.loads(raw)
        if not entries:
            return None

        first = entries[0]
        return Utxo(
            txid=first['tx_hash'],
            output_index=first['tx_pos'],
            satoshis=first['value'],
            script='',
        )

    def get_network(self) -> str:
        return self.network

    def get_raw_transaction(self, txid: str) -> str:
        raw = self._get(f'/tx/{txid}/hex')
        return raw.decode('utf-8').strip()

    def get_fee_rate(self) -> int:
        # BSV standard relay fee is 0.1 sat/byte (100 sat/KB).
        return 100

File: packages/runar-rb/lib/runar/sdk/woc_provider.rb
# frozen_string_literal: true

require 'net/http'
require 'json'
require 'uri'
require_relative 'provider'
require_relative 'types'

# WhatsOnChainProvider — HTTP-based BSV API provider.
#
# Implements the Provider interface by making HTTP requests to the
# WhatsOnChain REST API. Uses only Ruby stdlib (net/http, json, uri).
#
#   provider = Runar::SDK::WhatsOnChainProvider.new(network: 'mainnet')
#   utxos = provider.get_utxos('1A1zP1eP...')

module Runar
  module SDK
    class WhatsOnChainProvider < Provider
      # @param network [String] 'mainnet' or 'testnet'
      def initialize(network: 'mainnet')
        @network = network
        @base_url = if network == 'mainnet'
                      'https://api.whatsonchain.com/v1/bsv/main'
                    else
                      'https://api.whatsonchain.com/v1/bsv/test'
                    end
      end

      # Fetch a TransactionData by its txid.
      #
      # @param txid [String] transaction id (64 hex chars)
      # @return [TransactionData]
      def get_transaction(txid)
        data = api_get("/tx/hash/#{txid}")

        inputs = Array(data['vin']).map do |vin|
          TxInput.new(
            txid: vin['txid'],
            output_index: vin['vout'],
            script: vin.dig('scriptSig', 'hex') || '',
            sequence: vin['sequence'] || 0xFFFFFFFF
          )
        end

        outputs = Array(data['vout']).map do |vout|
          satoshis = (vout['value'].to_f * 1e8).round
          TxOutput.new(
            satoshis: satoshis,
            script: vout.dig('scriptPubKey', 'hex') || ''
          )
        end

        TransactionData.new(
          txid: data['txid'],
          version: data['version'] || 1,
          inputs: inputs,
          outputs: outputs,
          locktime: data['locktime'] || 0,
          raw: data['hex'] || ''
        )
      end

      # Broadcast a raw transaction hex to the network.
      #
      # @param raw_tx [String] hex-encoded raw transaction
      # @return [String] txid of the broadcasted transaction
      def broadcast(raw_tx)
        uri = URI("#{@base_url}/tx/raw")
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = true

        request = Net::HTTP::Post.new(uri)
        request['Content-Type'] = 'application/json'
        request.body = JSON.generate(txhex: raw_tx)

        response = http.request(request)
        unless response.is_a?(Net::HTTPSuccess)
          raise "WoC broadcast failed (#{response.code}): #{response.body}"
        end

        # WoC returns the txid as a JSON-encoded string.
        JSON.parse(response.body)
      end

      # Return all UTXOs for a given address.
      #
      # Note: WoC does not return locking scripts in UTXO lists, so the
      # script field is set to an empty string.
      #
      # @param address [String] BSV address
      # @return [Array<Utxo>]
      def get_utxos(address)
        entries = api_get("/address/#{address}/unspent")
        return [] unless entries.is_a?(Array)

        entries.map do |e|
          Utxo.new(
            txid: e['tx_hash'],
            output_index: e['tx_pos'],
            satoshis: e['value'],
            script: ''
          )
        end
      end

      # Find a contract UTXO by its script hash.
      #
      # @param script_hash [String] hex-encoded script hash
      # @return [Utxo, nil]
      def get_contract_utxo(script_hash)
        entries = api_get("/script/#{script_hash}/unspent")
        return nil unless entries.is_a?(Array) && !entries.empty?

        first = entries[0]
        Utxo.new(
          txid: first['tx_hash'],
          output_index: first['tx_pos'],
          satoshis: first['value'],
          script: ''
        )
      rescue RuntimeError => e
        # 404 simply means no UTXO found
        return nil if e.message.include?('404')

        raise
      end

      # Return the network this provider is connected to.
      #
      # @return [String] 'mainnet' or 'testnet'
      def get_network
        @network
      end

      # Fetch the raw transaction hex by its txid.
      #
      # @param txid [String]
      # @return [String] hex-encoded raw transaction
      def get_raw_transaction(txid)
        uri = URI("#{@base_url}/tx/#{txid}/hex")
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = true

        response = http.request(Net::HTTP::Get.new(uri))
        unless response.is_a?(Net::HTTPSuccess)
          raise "WoC getRawTransaction failed (#{response.code}): #{response.body}"
        end

        response.body.strip
      end

      # Return the current fee rate in satoshis per kilobyte.
      #
      # BSV standard relay fee is 0.1 sat/byte (100 sat/KB).
      #
      # @return [Numeric]
      def get_fee_rate
        100
      end

      private

      # Perform an HTTP GET request against the WoC API and parse the JSON response.
      #
      # @param path [String] API path (appended to base_url)
      # @return [Object] parsed JSON response
      def api_get(path)
        uri = URI("#{@base_url}#{path}")
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = true

        response = http.request(Net::HTTP::Get.new(uri))
        unless response.is_a?(Net::HTTPSuccess)
          raise "WoC request failed (#{response.code}): #{response.body}"
        end

        JSON.parse(response.body)
      end
    end
  end
end
